### ISSUE TITLE
genfromtxt with names=True doesn't work with python3

### ISSUE THREAD

**Issue by hsgg** [numpy/numpy#5411]
Using python 3.4.2 on Arch Linux with numpy 1.9.1, the following code doesn't work:

```
#!/usr/bin/env python

import sys
import numpy as np

data = np.genfromtxt(sys.stdin, names=True)
```

When running with a simple input file, it gives the following error:

```
Traceback (most recent call last):
  File "./test.py", line 6, in <module>
    data = np.genfromtxt(sys.stdin, names=True)
  File "/usr/lib/python3.4/site-packages/numpy/lib/npyio.py", line 1395, in genfromtxt
    if comments in first_line:
TypeError: 'in <string>' requires string as left operand, not bytes
```

The input file can be constructed with:

```python
f = io.StringIO(u"""\
x y z
2.4 2.3 0.1
3.5 5.6 0.2""")
```

There is no problem with python2.

**Comment by nigeil** [comment]
I've tested this a bit on my own arch linux install, and it seems that this has to do with how python3 handles bytestrings and your use of `sys.stdin` to request filenames. From the comments for genfromtxt: 
```
fname : file or str
File, filename, or generator to read. If the filename extension is
`.gz` or `.bz2`, the file is first decompressed. Note that
generators must return byte strings in Python 3k.
```
"**Note that generators must return byte strings in Python 3k.**" I suspect that `sys.stdin` is not returning the byte string that python3 requires, but I'm not familiar enough with it to know why.

**Solution**
Replace `sys.stdin` with `input`, which is a builtin function for python3 (`raw_input` for python2). This should work almost identically to your current implementation.
```
data = np.genfromtxt(input(), names=True)
```
Or for a more interactive experience, you can have it print a string to the console first:
```
data = np.genfromtxt(input("Please input a filename here: "), names=True)
```

**Comment by hsgg** [comment]
Thanks @nigeil for your comment, but that solves a different problem. I want to call my program as

    $ ./test.py < test.dat

That way I don't need to bother with the filename in python.
I believe you are right that sys.stdin returns strings, whereas genfromtxt requires bytestrings. I will need to read up on how to get a file-like object that returns bytestrings.

**Comment by nigeil** [comment]
Ah, I see the difference. Darned python3. You could replicate the functionality you're looking for with my solution by instead passing a file containing only the filename of the data file - but that is pretty lame. If you find a solution, please share it!

**Comment by jaimefrio** [comment]
With Python 3.3.2 and numpy 1.8, I can reproduce your error, and get it to work using `sys.stdin.buffer` instead of `sys.stdin`, you may want to give it a try.

**Comment by hsgg** [comment]
Thanks @jaimefrio. `sys.stdin.buffer` does indeed work. Personally, I think `genfromtxt()` should support `sys.stdin` directly, as it feels more natural to me. What do you think?

**Comment by jaimefrio** [comment]
Yes, I think I agree: it seems like an important shortcoming in Python 3 support. That said, I don't know that code enough to understand what the implications would be.

My general feeling is that numpy's read-from-file functionality needs a serious overhaul to bring it up to the standard that pandas has set, and I am not sure that small incremental changes will take us there.

 But PRs are always welcome! ;-)

**Comment by eric-wieser** [comment]
Fixed in 1.14
